File: examples/python/consulta.py
"""MyCEP — Python 3.9+.

    MYCEP_API_KEY=cep_live_... python consulta.py 01001000

Única dependência: requests.
"""
from __future__ import annotations

import os
import re
import sys

import requests

BASE = os.environ.get("MYCEP_BASE", "https://api.mycep.app.br")


class MyCepError(Exception):
    def __init__(self, status: int, code: str, message: str) -> None:
        super().__init__(message)
        self.status = status
        self.code = code


def buscar_cep(cep: str) -> dict:
    digitos = re.sub(r"\D", "", cep)
    if len(digitos) != 8:
        raise MyCepError(400, "invalid_cep", "CEP precisa de 8 dígitos.")

    headers = {"Accept": "application/json"}
    chave = os.environ.get("MYCEP_API_KEY")
    if chave:
        headers["Authorization"] = f"Bearer {chave}"

    resposta = requests.get(
        f"{BASE}/api/v1/search/cep/{digitos}", headers=headers, timeout=10
    )
    if resposta.status_code != 200:
        corpo = {}
        try:
            corpo = resposta.json()
        except ValueError:
            pass
        raise MyCepError(
            resposta.status_code,
            corpo.get("error", "request_failed"),
            corpo.get("message", "Falha na consulta."),
        )
    return resposta.json()


if __name__ == "__main__":
    try:
        endereco = buscar_cep(sys.argv[1] if len(sys.argv) > 1 else "01001000")
    except MyCepError as erro:
        print(f"[{erro.code}] {erro}", file=sys.stderr)
        raise SystemExit(1)
    print(f"{endereco['logradouro']} — {endereco['cidade']}/{endereco['uf']}")

File: examples/typescript/consulta.ts
/**
 * MyCEP — TypeScript / Node 20+.
 *   MYCEP_API_KEY=cep_live_... npx tsx consulta.ts 01001000
 *
 * Sem dependência: usa o `fetch` nativo. O SDK oficial (../../sdk/typescript)
 * faz o mesmo com tipos prontos.
 */
const BASE = process.env.MYCEP_BASE ?? 'https://api.mycep.app.br'

export interface Endereco {
  cep: string
  tipoLogradouro: string | null
  logradouro: string
  bairro: string | null
  cidade: string
  uf: string
  ibge: string
  latitude?: number
  longitude?: number
}

export class MyCepError extends Error {
  constructor(
    readonly status: number,
    readonly code: string,
    message: string
  ) {
    super(message)
  }
}

export async function buscarCep(cep: string): Promise<Endereco> {
  const digitos = cep.replace(/\D/g, '')
  if (digitos.length !== 8) throw new MyCepError(400, 'invalid_cep', 'CEP precisa de 8 dígitos.')

  const headers: Record<string, string> = { accept: 'application/json' }
  if (process.env.MYCEP_API_KEY) {
    headers.authorization = `Bearer ${process.env.MYCEP_API_KEY}`
  }

  const resposta = await fetch(`${BASE}/api/v1/search/cep/${digitos}`, {
    headers,
    signal: AbortSignal.timeout(10_000),
  })

  if (!resposta.ok) {
    const corpo = (await resposta.json().catch(() => ({}))) as { error?: string; message?: string }
    throw new MyCepError(
      resposta.status,
      corpo.error ?? 'request_failed',
      corpo.message ?? 'Falha na consulta.'
    )
  }
  return (await resposta.json()) as Endereco
}

if (import.meta.url === `file://${process.argv[1]}`) {
  buscarCep(process.argv[2] ?? '01001000')
    .then((e) => console.log(`${e.logradouro} — ${e.cidade}/${e.uf} (IBGE ${e.ibge})`))
    .catch((e: MyCepError) => {
      console.error(`[${e.code}] ${e.message}`)
      process.exit(1)
    })
}
